File: pdk-ts/src/commands/assets.ts
/**
 * `pdk-ts assets` — sign Assets pallet operations.
 *
 * History, because this file used to claim otherwise: @polkadot/api signed
 * these calls correctly from day one, while Python `substrate-interface`
 * could not sign them at all on Portaldot's V13 metadata — `Assets.create`
 * was rejected at the RPC layer with "Invalid Transaction: bad signature"
 * before reaching dispatch. That gap is closed as of pdk 0.2.0: the Python
 * side root-caused it (a bare `Balance` resolving to u128 where
 * pallet-assets uses u64) and patched both the call encoding and the event
 * decode, so `pdk assets` now exists too. The pair is at parity here;
 * pick whichever runtime your project already has.
 *
 * Three lifecycle operations — the minimum to create, fund, and move a
 * custom asset: `create`, `mint`, `transfer`. All three submit through
 * send.ts's shared `submitExtrinsic` engine, so they inherit its
 * false-success guard and hang guard for free — no new signing-safety
 * code to get wrong here.
 */

import pc from 'picocolors';
import {Keyring} from '@polkadot/api';
import type {ApiPromise} from '@polkadot/api';
import {getApi, closeApi} from '../core/chain.js';
import {resolveNode} from '../core/config.js';
import {humanizeChainError} from '../core/errors.js';
import {normaliseUri} from './keys.js';
import {resolveRecipient, submitExtrinsic, type TransferOutcome} from './send.js';

export interface AssetsCommonOptions {
  from?: string;
  node?: string;
  timeout?: string;
  json?: boolean;
}

export interface AssetsCreateOptions extends AssetsCommonOptions {
  minBalance?: string;
  admin?: string;
}

export interface AssetsAmountOptions extends AssetsCommonOptions {
  amount?: string;
}

function parseAssetId(raw: string): number {
  // Number('') is 0, not NaN — require a non-empty digit string explicitly
  // so a blank/whitespace id doesn't silently resolve to asset #0.
  if (!/^\d+$/.test(raw.trim())) throw new Error(`invalid asset id "${raw}" — must be a non-negative integer.`);
  return Number(raw.trim());
}

/** Plain non-negative integer amount — Assets pallet units are whole (no fixed decimals convention like POT). */
function parseAssetAmount(raw: string | undefined): bigint {
  if (raw === undefined || !/^\d+$/.test(raw.trim())) throw new Error('Provide a non-negative integer --amount.');
  return BigInt(raw.trim());
}

function emitOutcome(action: string, extra: Record<string, unknown>, outcome: TransferOutcome, json: boolean): boolean {
  const ok = outcome.status === 'success';
  if (json) {
    console.log(JSON.stringify({success: ok, status: outcome.status, tx: outcome.txHash, block: outcome.block, error: outcome.error, ...extra}));
    return ok;
  }
  if (outcome.status === 'success') {
    console.log(pc.green(`\n  ✓ ${action}`));
    console.log(pc.dim(`  tx: ${outcome.txHash}\n`));
  } else if (outcome.status === 'unconfirmed') {
    console.error(pc.yellow(`\n  ⚠ submitted but the outcome couldn't be confirmed — @polkadot/api can't decode Portaldot's result events.`));
    console.error(pc.dim(outcome.txHash ? `  Confirm with Python: pdk debug ${outcome.txHash}\n` : '  No block inclusion within the timeout — check node connectivity and retry.\n'));
  } else {
    console.error(pc.red(`\n  ✗ ${action} failed${outcome.error ? ` — ${outcome.error}` : ''}`));
    console.error(pc.dim(`  tx: ${outcome.txHash}\n`));
  }
  return ok;
}

async function withApi<T>(node: string | undefined, timeout: string | undefined, fn: (api: ApiPromise, keyring: Keyring) => Promise<T>): Promise<T> {
  const api = await getApi(resolveNode(node), timeout ? Math.round(Number(timeout) * 1000) : undefined);
  const keyring = new Keyring({type: 'sr25519', ss58Format: 42});
  return fn(api, keyring);
}

export async function runCreate(id: string, opts: AssetsCreateOptions): Promise<void> {
  const node = resolveNode(opts.node);
  try {
    const assetId = parseAssetId(id);
    // Same strict digit-only guard as parseAssetAmount — BigInt() alone
    // accepts "" (→0n) and hex like "0x10" (→16n), silently corrupting the
    // min-balance a user typed. Verified live: --min-balance "0x10" created
    // an asset with min_balance 16, not 10, with no error.
    const minBalance = opts.minBalance !== undefined ? parseAssetAmount(opts.minBalance) : 1n;
    const outcome = await withApi(opts.node, opts.timeout, async (api, keyring) => {
      const admin = keyring.addFromUri(normaliseUri(opts.from ?? '//Alice'));
      const adminAddress = opts.admin ? resolveRecipient(keyring, opts.admin) : admin.address;
      const tx = api.tx.assets.create(assetId, adminAddress, minBalance.toString());
      return submitExtrinsic(api, tx, admin);
    });
    const ok = emitOutcome(`created asset #${id}`, {id: Number(id)}, outcome, Boolean(opts.json));
    await closeApi();
    if (!ok) process.exit(1);
  } catch (err) {
    await fail('assets create', err, node, opts.json);
  }
}

export async function runMint(id: string, to: string, opts: AssetsAmountOptions): Promise<void> {
  const node = resolveNode(opts.node);
  try {
    const assetId = parseAssetId(id);
    const amount = parseAssetAmount(opts.amount);
    const outcome = await withApi(opts.node, opts.timeout, async (api, keyring) => {
      const admin = keyring.addFromUri(normaliseUri(opts.from ?? '//Alice'));
      const dest = resolveRecipient(keyring, to);
      const tx = api.tx.assets.mint(assetId, dest, amount.toString());
      return submitExtrinsic(api, tx, admin);
    });
    const ok = emitOutcome(`minted ${opts.amount} of asset #${id} to ${to}`, {id: Number(id), to, amount: opts.amount}, outcome, Boolean(opts.json));
    await closeApi();
    if (!ok) process.exit(1);
  } catch (err) {
    await fail('assets mint', err, node, opts.json);
  }
}

export async function runTransfer(id: string, to: string, opts: AssetsAmountOptions): Promise<void> {
  const node = resolveNode(opts.node);
  try {
    const assetId = parseAssetId(id);
    const amount = parseAssetAmount(opts.amount);
    const outcome = await withApi(opts.node, opts.timeout, async (api, keyring) => {
      const sender = keyring.addFromUri(normaliseUri(opts.from ?? '//Alice'));
      const dest = resolveRecipient(keyring, to);
      const tx = api.tx.assets.transfer(assetId, dest, amount.toString());
      return submitExtrinsic(api, tx, sender);
    });
    const ok = emitOutcome(`transferred ${opts.amount} of asset #${id} to ${to}`, {id: Number(id), to, amount: opts.amount}, outcome, Boolean(opts.json));
    await closeApi();
    if (!ok) process.exit(1);
  } catch (err) {
    await fail('assets transfer', err, node, opts.json);
  }
}

async function fail(action: string, err: unknown, node: string, json?: boolean): Promise<void> {
  const msg = humanizeChainError(err, node);
  if (json) console.log(JSON.stringify({error: msg, endpoint: node}));
  else console.error(pc.red(`\n  ✗ ${action} failed — ${msg}\n`));
  await closeApi();
  process.exit(1);
}

File: pdk/commands/assets.py
"""`pdk assets` — create, mint, and transfer custom assets.

Three lifecycle operations, the minimum to create an asset, fund it, and
move it: ``create``, ``mint``, ``transfer``. Mirrors ``pdk-ts assets``
argument-for-argument, closing the last command-surface gap between the
two CLIs.

Signing these on Portaldot needed two separate fixes, both of which live
in ``pdk/core`` rather than here:

* ``chain._fix_assets_balance_width`` — ``Assets.create``'s ``min_balance``
  is declared ``T::Balance`` in V13 metadata, which substrate-interface
  resolves as u128 while pallet-assets uses u64. The extra 8 bytes get
  signed but not verified, so the node rejects the extrinsic with
  "bad signature" before dispatch.
* ``events.receipt_succeeded`` — the same ambiguity on the event side
  makes ``receipt.is_success`` raise for ``mint``/``transfer``, because a
  block's events decode as one vec and an over-wide read desynchronises
  the rest of it.

So this module submits through ``submit_call`` and confirms through
``receipt_succeeded``, and inherits both. Success is confirmed only by a
positive ``System.ExtrinsicSuccess`` for this extrinsic — never by the
absence of an error.
"""

from __future__ import annotations

import json as jsonlib

import typer
from rich.console import Console
from rich.markup import escape
from substrateinterface import Keypair

from pdk.config import DEFAULT_NODE_URL
from pdk.core.chain import connect, normalise_account_uri, resolve_account, submit_call
from pdk.core.decoder import decode_receipt
from pdk.core.events import receipt_succeeded

console = Console()

app = typer.Typer(
    name="assets",
    help="Create, mint, and transfer custom assets (Assets pallet).",
    no_args_is_help=True,
    add_completion=False,
)

_NODE = typer.Option(DEFAULT_NODE_URL, "--node", help="Portaldot node WS endpoint.")
_FROM = typer.Option("//Alice", "--from", help="Signing account URI.")
_JSON = typer.Option(False, "--json", help="Emit machine-readable JSON.")


def _parse_amount(raw: str, label: str) -> int:
    """Whole non-negative integer only.

    Assets pallet units have no decimals convention like POT, so there is
    nothing to scale. int() alone would accept "0x10" (→16) and a bare
    "-1", either of which silently encodes something other than what was
    typed — refuse instead.
    """
    text = str(raw).strip()
    if not text.isdigit():
        raise typer.BadParameter(f"{label} must be a non-negative whole number, got {raw!r}.")
    return int(text)


def _connect(node: str, json_out: bool):
    try:
        return connect(node)
    except Exception as exc:  # noqa: BLE001
        if json_out:
            typer.echo(jsonlib.dumps({"error": f"Cannot reach a Portaldot node at {node}", "detail": str(exc)}))
        else:
            console.print(f"[red]Cannot reach a Portaldot node at {node}[/red]")
            console.print(f"[dim]{exc}[/dim]")
            console.print("Start a node with [bold]pdk up[/bold] (run the node in WSL on Windows; pdk itself runs natively).")
        raise typer.Exit(code=1)


def _submit(substrate, sender: str, call: str, params: dict, action: str,
            extra: dict, json_out: bool) -> None:
    """Sign, submit, confirm, and report one Assets call."""
    keypair = Keypair.create_from_uri(normalise_account_uri(sender))
    try:
        receipt = submit_call(substrate, keypair, "Assets", call, params)
    except Exception as exc:  # noqa: BLE001
        if json_out:
            typer.echo(jsonlib.dumps({"success": False, "error": str(exc), **extra}))
        else:
            console.print(f"[red]✗ assets {call} failed: {escape(str(exc))}[/red]")
        raise typer.Exit(code=1)

    tx = str(receipt.extrinsic_hash)
    if receipt_succeeded(substrate, receipt):
        if json_out:
            typer.echo(jsonlib.dumps({"success": True, "tx": tx, "block": receipt.block_hash, **extra}))
        else:
            console.print(f"[green]✓ {action}[/green]")
            console.print(f"[dim]tx: {tx}[/dim]")
        return

    decoded = decode_receipt(receipt, substrate)
    name = decoded.name if decoded else "unknown"
    if json_out:
        typer.echo(jsonlib.dumps({"success": False, "tx": tx, "error": name, **extra}))
    else:
        console.print(f"[red]✗ assets {call} failed: {escape(name)}[/red]")
        console.print(f"[dim]Diagnose it: pdk debug {tx}[/dim]")
    raise typer.Exit(code=1)


@app.command("create")
def create(
    asset_id: str = typer.Argument(..., help="Numeric id for the new asset."),
    min_balance: str = typer.Option("1", "--min-balance", help="Smallest holdable amount; accounts below it are dusted."),
    admin: str = typer.Option(None, "--admin", help="Admin account (//URI or SS58). Defaults to the signer."),
    sender: str = _FROM,
    node: str = _NODE,
    json_out: bool = _JSON,
) -> None:
    """Create a new asset, owned and administered by --admin (default: the signer)."""
    ident = _parse_amount(asset_id, "asset id")
    minimum = _parse_amount(min_balance, "--min-balance")
    substrate = _connect(node, json_out)
    admin_address = resolve_account(admin) if admin else Keypair.create_from_uri(normalise_account_uri(sender)).ss58_address
    _submit(
        substrate, sender, "create",
        {"id": ident, "admin": admin_address, "min_balance": minimum},
        f"created asset #{ident}", {"id": ident, "admin": admin_address}, json_out,
    )


@app.command("mint")
def mint(
    asset_id: str = typer.Argument(..., help="Asset id to mint."),
    to: str = typer.Argument(..., help="Beneficiary (//URI or SS58)."),
    amount: str = typer.Option(..., "--amount", help="How much to mint."),
    sender: str = _FROM,
    node: str = _NODE,
    json_out: bool = _JSON,
) -> None:
    """Mint new units of an asset to an account. Only the asset's issuer may do this."""
    ident = _parse_amount(asset_id, "asset id")
    units = _parse_amount(amount, "--amount")
    substrate = _connect(node, json_out)
    dest = resolve_account(to)
    _submit(
        substrate, sender, "mint",
        {"id": ident, "beneficiary": dest, "amount": units},
        f"minted {units:,} of asset #{ident} to {to}",
        {"id": ident, "to": dest, "amount": units}, json_out,
    )


@app.command("transfer")
def transfer(
    asset_id: str = typer.Argument(..., help="Asset id to transfer."),
    to: str = typer.Argument(..., help="Recipient (//URI or SS58)."),
    amount: str = typer.Option(..., "--amount", help="How much to transfer."),
    sender: str = _FROM,
    node: str = _NODE,
    json_out: bool = _JSON,
) -> None:
    """Transfer units of an asset from the signer to another account."""
    ident = _parse_amount(asset_id, "asset id")
    units = _parse_amount(amount, "--amount")
    substrate = _connect(node, json_out)
    dest = resolve_account(to)
    _submit(
        substrate, sender, "transfer",
        {"id": ident, "target": dest, "amount": units},
        f"transferred {units:,} of asset #{ident} to {to}",
        {"id": ident, "to": dest, "amount": units}, json_out,
    )
